### ISSUE TITLE
Remote code execution vulnerability in NLTK

### ISSUE THREAD

**Issue by Dunedan** [nltk/nltk#3266]
The current and earlier versions of NLTK are vulnerable to a remote code execution vulnerability when using the integrated data package download functionality. A man-in-the-middle attacker or an attacker with control over the NLTK data index can force users which use data packages with pickled Python code to download a new version of the package which executes arbitrary code upon unpickling.

Data packages which have been identified to be vulnerable are "averaged_perceptron_tagger" and "punkt". For code to be vulnerable it has to download a data package and use functionality in NLTK which causes the data package to be unpickled.

Here is an example of vulnerable code for the "averaged_perceptron_tagger" data package:

```python
import nltk
nltk.download("averaged_perceptron_tagger")
nltk.pos_tag(["hello", "world"])
```

This vulnerability was reported together with POC code to exploit it multiple times to the NLTK team via the email address mentioned in [SECURITY.md](SECURITY.md) (and later directly to some maintainers as well). It was reported first on 2024-05-19. So far there has been no response from the NLTK maintainers to these reports.

https://github.com/nltk/nltk/issues/2522 already pointed out the security implications of using pickled code a few years ago, but didn't receive a response either.

**Comment by ekaf** [comment]
@Dunedan, what would you suggest?  
Maybe print a warning and ask for confirmation before loading a pickle?

The pickles in question contain Python classes with executable functions. They are not data but programs. And,as a compiled format, they are not ideal for an open-source project. 

But an eventual attacker needs to trick the user into loading a malicious pickle.

**Comment by Dunedan** [comment]
The short answer to solve this issue is to remove the download of code via network at runtime.

If that's not feasible for NLTK I have no clear answer how to solve this issue. However, there are some measures I'd implement, which would at least improve the situation:

#### Remove use of pickled code wherever possible

While I haven't checked what the code in in the data packages with pickled code does, I can imagine that it's not the code which is large, but rather data associated with it. Maybe the algorithms implemented in there can be moved into NLTK, removing the need for pickled code.

Also instead of using pickle, a domain-specific format could be used (depending on the use case something like ONNX (https://onnx.ai/) might fit for that).

#### Make pickled code reproducible

When an update for a data package with pickled code is issued, it's difficult to figure out what changed. Having the pickle files be reproducibly built would be helpful for assessing if they're legit.

#### Make documentation clear for download once during setup

While the documentation details various ways to install data packages right now, making it more prominent that data packages should be installed once during installation/setup, instead of including `Downloader.download()` in application code, could be helpful to make users aware of the implications downloading data packages in their application code has.

#### Opt-in for auto-updating installed data packages

Right now calling `Downloader().download()` always results in the data index being fetched and potential updates to already installed packages being downloaded and installed. Instead I believe the download functionality should by default only do network requests and install data packages if they're not already installed. Automatic updating could then be an additional parameter users have to explicitly set.

#### Opt-in for installing data packages containing code

Similar to having users explicitly opt-in for auto-updates, I'd do the same for downloads of data packages which contain code which is going to be executed.

In addition I'd always log a warning when a package with pickled code gets downloaded or used.

#### Verify downloaded packages

Right now the data index contains a checksum per package. However this checksum is only used to check whether a downloaded package differs from a previously downloaded version. It is not used right now to check if a downloaded package is the one referenced by the data index. This should be done together with the move to a cryptographical strong hash function for the checksums.

**Comment by ekaf** [comment]
Thanks @dunedan for your detailed suggestions.

Should this issue be labelled critical? I wonder if it could make NLTK deemed unsafe for use in schools, or inclusion in some software distributions. But on the other hand, has any exploit been reported since #2522, back in 2020? When nobody else than the administrator controls nltk_data, the pickles should be safe to load.

It could be nice to remove all the pickles completely, for ex. by moving the code into the classes of the corresponding corpus reader in NLTK, and reducing the data package, for ex. to CSV files for tabular data, and JSON or XML for complex data.

**Comment by Dunedan** [comment]
This is now being tracked as CVE-2024-39705.

**Comment by ekaf** [comment]
In addition to the two packages already mentioned, the following also contain pickles:

- chunkers/maxent_ne_chunker.zip
- help/tagsets.zip
- taggers/maxent_treebank_pos_tagger.zip

In total nltk_data contains 52 pickles, where half are Python 2, which is not supported anymore, so these may just be deleted.

Starting at the easy end, the pickles in _help/tagsets.zip_ are very easy to decompile, since they are just Python dictionaries that associate each tag symbol with a tuple consisting in the tag name, and a string of instances. This is just tabular data, that would fit very well in the CSV or TAB format. One may wonder what made it seem attractive to compile this data into pickles back then.

**Comment by DhenPadilla** [comment]
Hey team!!!

Thanks for raising this @Dunedan and for the insightful thread @ekaf!

We hit a critical dependabot alert regarding this vulnerability. So I'm now tracking this thread quite closely!
I was trying to trace down into its sub dependencies but got super lost with where to be looking in order to verify that these vulnerable data packages are being loaded into our codebase;

Sorry for the silly questions here; but would love to understand this a little more;

> A man-in-the-middle attacker or an attacker with control over the NLTK data index can force users which use data packages with pickled Python code to download a new version of the package which executes arbitrary code upon unpickling.

1. This is the first time I've heard of `pickling`! So thank you for raising this and introducing me to this new concept for me.
So, my understanding of the vulnerability here is that the pickled byte-stream can be manipulated before unpickling the content which may include malicious executable code?

2. We're currently using the `LazyCorpusLoader` and we download the `words` package. My understanding here is that the `download()` function is a main vulnerable entry-point here? 
To add fuel to this fire; we use this within an AWS lambda function which means we would run a `download` every single run.. making this really open :/ - What's the best work around/fix here for us (if needed)?

3. How can I tell whether the `words` package is a pickled download? Or for a broader question; what's the best way to trace which datasets are prone to this vulnerability?

**Comment by ekaf** [comment]
```
#!/bin/sh
for f in ~/nltk_data/*/*zip
do
unzip -l $f | grep -i pickle >> Nltk_Pickles.txt
done
```
[Nltk_Pickles.txt](https://github.com/user-attachments/files/16054731/Nltk_Pickles.txt)

**Comment by mcepl** [comment]
On Mon Jul 1, 2024 at 1:35 PM CEST, DhenPadilla wrote:
> 1. This is the first time I've heard of `pickling`! So thank you for raising this and introducing me to this new concept for me.
> So, my understanding of the vulnerability here is that the pickled byte-stream can be manipulated before unpickling the content which may include malicious executable code?

See https://docs.python.org/3/library/pickle.html and big red warning there. Pickle can contain ANY Python data including but not limited to executable functions. So, pickle should NEVER be used for data which are outside of your control. Especially, downloading pickles from Internet carries with itself serious risk of downloading and executing malicious code.

**Comment by mcepl** [comment]
I understand this is just an emergency maintenance brutal workaround, but would anything outside of the downloading stuff from Internet break with this patch?

```diff
---
 nltk/app/chartparser_app.py    |   13 +++++++++++++
 nltk/corpus/reader/util.py     |    2 ++
 nltk/data.py                   |    2 ++
 nltk/parse/transitionparser.py |    2 ++
 nltk/tbl/demo.py               |    4 +++-
 5 files changed, 22 insertions(+), 1 deletion(-)

--- a/nltk/app/chartparser_app.py
+++ b/nltk/app/chartparser_app.py
@@ -800,6 +800,10 @@ class ChartComparer:
             showerror("Error Saving Chart", f"Unable to open file: {filename!r}\n{e}")
 
     def load_chart_dialog(self, *args):
+        showerror("Security Error",
+                  "Due to gh#nltk/nltk#3266, deserializing from " +
+                  "a pickle is forbidden.")
+        return
         filename = askopenfilename(
             filetypes=self.CHART_FILE_TYPES, defaultextension=".pickle"
         )
@@ -811,6 +815,8 @@ class ChartComparer:
             showerror("Error Loading Chart", f"Unable to open file: {filename!r}\n{e}")
 
     def load_chart(self, filename):
+        raise RuntimeError("Due to gh#nltk/nltk#3266, deserializing from " +
+                           "a pickle is forbidden.")
         with open(filename, "rb") as infile:
             chart = pickle.load(infile)
         name = os.path.basename(filename)
@@ -2268,6 +2274,10 @@ class ChartParserApp:
         if not filename:
             return
         try:
+            showerror("Security Error",
+                      "Due to gh#nltk/nltk#3266, deserializing from " +
+                      "a pickle is forbidden.")
+            return
             with open(filename, "rb") as infile:
                 chart = pickle.load(infile)
             self._chart = chart
@@ -2306,6 +2316,9 @@ class ChartParserApp:
             return
         try:
             if filename.endswith(".pickle"):
+                showerror("Due to gh#nltk/nltk#3266, deserializing from " +
+                          "a pickle is forbidden.")
+                return
                 with open(filename, "rb") as infile:
                     grammar = pickle.load(infile)
             else:
--- a/nltk/corpus/reader/util.py
+++ b/nltk/corpus/reader/util.py
@@ -521,6 +521,8 @@ class PickleCorpusView(StreamBackedCorpu
 
     def read_block(self, stream):
         result = []
+        raise RuntimeError("Due to gh#nltk/nltk#3266, deserializing from " +
+                           "a pickle is forbidden.")
         for i in range(self.BLOCK_SIZE):
             try:
                 result.append(pickle.load(stream))
--- a/nltk/data.py
+++ b/nltk/data.py
@@ -752,6 +752,8 @@ def load(
     if format == "raw":
         resource_val = opened_resource.read()
     elif format == "pickle":
+        raise RuntimeError("Due to gh#nltk/nltk#3266, deserializing from " +
+                           "a pickle is forbidden.")
         resource_val = pickle.load(opened_resource)
     elif format == "json":
         import json
--- a/nltk/parse/transitionparser.py
+++ b/nltk/parse/transitionparser.py
@@ -553,6 +553,8 @@ class TransitionParser(ParserI):
         """
         result = []
         # First load the model
+        raise RuntimeError("Due to gh#nltk/nltk#3266, deserializing from " +
+                           "a pickle is forbidden.")
         model = pickle.load(open(modelFile, "rb"))
         operation = Transition(self._algorithm)
 
--- a/nltk/tbl/demo.py
+++ b/nltk/tbl/demo.py
@@ -253,6 +253,8 @@ def postag(
                 )
             )
         with open(cache_baseline_tagger) as print_rules:
+            raise RuntimeError("Due to gh#nltk/nltk#3266, deserializing from " +
+                               "a pickle is forbidden.")
             baseline_tagger = pickle.load(print_rules)
             print(f"Reloaded pickled tagger from {cache_baseline_tagger}")
     else:
@@ -327,7 +329,7 @@ def postag(
         with open(serialize_output) as print_rules:
             brill_tagger_reloaded = pickle.load(print_rules)
         print(f"Reloaded pickled tagger from {serialize_output}")
-        taggedtest_reloaded = brill_tagger.tag_sents(testing_data)
+        taggedtest_reloaded = brill_tagger_reloaded.tag_sents(testing_data)
         if taggedtest == taggedtest_reloaded:
             print("Reloaded tagger tried on test set, results identical")
         else:
```

**Comment by cassneal** [comment]
Hello, is there an ETA for a fix?

**Comment by ekaf** [comment]
Before getting too alarmed, we may want to wait for a sober analysis of this vulnerability, bearing in mind that it has been known for several years, without any known exploitation yet.

Still, pickles are not open source, which is in itself a good reason to avoid them. In addition to repackaging the data, a complete fix would require deep modifications in several Python files:
```

#!/bin/sh

tree=nltk
q=pickl
of="$tree"_$q.txt
echo -n > $of 
mxd=`ls -R $tree|grep /|awk -F "/" '{print NF}'|sort -nr|head -1`
 
for f in `find -L $tree -maxdepth $mxd -name "*py"`
do
if [ -n "`grep pickle $f`" ]
then
echo $f >> $of 
echo -------------------------------- >> $of
grep -ni $q $f >> $of
echo \\n--------------------------------------------------------------------- >> $of
fi
done

```

[nltk_pickl.txt](https://github.com/user-attachments/files/16081709/nltk_pickl.txt)

**Comment by nicolaschaillan** [comment]
We really need a fix on this. Lots of customers are fricking out when they see a High CVE finding... Any ETA?

**Comment by nicolaschaillan** [comment]
> Before getting too alarmed, we may want to wait for a sober analysis of this vulnerability, bearing in mind that it has been known for several years, without any known exploitation yet.
> 
> Still, pickles are not open source, which is in itself a good reason to avoid them. In addition to repackaging the data, a complete fix would require deep modifications in several Python files:
> 
> ```
> 
> #!/bin/sh
> 
> tree=nltk
> q=pickl
> of="$tree"_$q.txt
> echo -n > $of 
> mxd=`ls -R $tree|grep /|awk -F "/" '{print NF}'|sort -nr|head -1`
>  
> for f in `find -L $tree -maxdepth $mxd -name "*py"`
> do
> if [ -n "`grep pickle $f`" ]
> then
> echo $f >> $of 
> echo -------------------------------- >> $of
> grep -ni $q $f >> $of
> echo \\n--------------------------------------------------------------------- >> $of
> fi
> done
> ```
> 
> [nltk_pickl.txt](https://github.com/user-attachments/files/16081709/nltk_pickl.txt)

Now that it has a public NIST CVE, people will exploit it. This isn't something you "wait" on. This is something you address right away...

**Comment by nicolaschaillan** [comment]
> ```diff
> ```diff
> (testing_data)
> +        taggedtest_reloaded = brill_tagger_reloaded.tag_sents(testing_data)
>          if taggedtest == taggedtest_reloaded:
>              print("Reloaded tagger tried on test set, results identical")
>          else:
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
>       
> ```

Could you share the .patch file by any chance? getting character issues when trying to apply..

**Comment by adivekar-utexas** [comment]
+1 we need a fix on this...many repos on https://github.com/amazon-science/ use `nltk` and are affected.

**Comment by Dunedan** [comment]
A possible mitigation is to not use NLTK's downloader functionality to download vulnerable data packages until this issue is fixed. Instead you could download the data packages your application needs manually once, verify they are legit and ship them bundled with your application.

**Comment by wallies** [comment]
As @Dunedan mentioned above. A workaround for now is to consider making your own package index.
Considering the package index that is used is https://github.com/nltk/nltk_data. You can create your own fork and validate the packages you need and customise the mirror used by using --url

**Comment by ekaf** [comment]
> Now that it has a public NIST CVE, people will exploit it. This isn't something you "wait" on. This is something you address right away...

@nicolaschaillan,  before even starting NLTK, you load Python, which already exposes the user to the risk of being tricked into downloading malicious pickles anyway. In that context, wouldn't a fix in NLTK only marginally improve the user's security?

**Comment by nicolaschaillan** [comment]
> > Now that it has a public NIST CVE, people will exploit it. This isn't something you "wait" on. This is something you address right away...
> 
> @nicolaschaillan, before even starting NLTK, you load Python, which already exposes the user to the risk of being tricked into downloading malicious pickles anyway. In that context, wouldn't a fix in NLTK only marginally improve the user's security?

this isn't a discussion to see if there are other ways to get into something. Python has no such finding open at this time. This is a discussion to fix this CVE. 
This feature should be DISABLED unless the administrator explicitly enables it, with a narrow whitelist of resources. 

At the end of the day, while I understand this could seem like an "expected behavior" and you can go fight that fight with NVD, you should have an option to enable this with basic ENV variable (disabled by default) or something like this.

Not sure why your contributors are making all this more difficult than it got to be. a CVE is a big deal for enterprises. You AI people seem not to understand cyber 101.

**Comment by adivekar-utexas** [comment]
@ekaf I agree with @nicolaschaillan and let me explain that I've gotten a high-severity ticket (both internally at Amazon and [from GitHub's dependabot](https://github.com/amazon-science/synthesizrr/security/dependabot/6)) because of this vulnerability. 

Given the options of (a) not resolving this ticket and (b) removing all dependencies on `nltk` from my codebase, I am **forced** to choose (b), in order to comply with my organization's security requirements. It's either that or take down my repository.

**Comment by ekaf** [comment]
I inspected a typical pickle in each series, looking at their type() and eventual \_\_dict\_\_.

The _tagsets_ and _averaged\_perceptron\_tagger_ packages contain only simple data structures, and can easily be translated to a text-based data format, which solves the problem radically.

The rest of the pickles contain complex classes. A good translation may only be possible with help from the original package authors, and would require more time.

So at this moment, users who find themselves in an emergency, and need very strict settings, may need to consider some of the mitigations suggested above, like [forbidding unpickling](https://github.com/nltk/nltk/issues/3266#issuecomment-2204404662), or using [their own secure fork](https://github.com/nltk/nltk/issues/3266#issuecomment-2208181098).

**Comment by nicolaschaillan** [comment]
> I inspected a typical pickle in each series, looking at their type() and eventual __dict__.
> 
> The _tagsets_ and _averaged_perceptron_tagger_ packages contain only simple data structures, and can easily be translated to a text-based data format, which solves the problem radically.
> 
> The rest of the pickles contain complex classes. A good translation may only be possible with help from the original package authors, and would require more time.
> 
> So at this moment, users who find themselves in an emergency, and need very strict settings, may need to consider some of the mitigations suggested above, like [forbidding unpickling](https://github.com/nltk/nltk/issues/3266#issuecomment-2204404662), or using [their own secure fork](https://github.com/nltk/nltk/issues/3266#issuecomment-2208181098).

The solution is simple. Whitelist which pickles can be loaded. Give control to the tenant to decide which ones are fine or not. Maybe by name or something. Right now, the fear is that third party can manage to load some that are malicious.

**Comment by mcepl** [comment]
> The solution is simple. Whitelist which pickles can be loaded. Give control to the tenant to decide which ones are fine or not. Maybe by name or something. Right now, the fear is that third party can manage to load some that are malicious.

I don’t think, that’s good enough. As long as there is an opportunity how to make NLTK download pickle from the Internet (by using some internal undocumented functions, there are really no private and really inaccessible functions in Python), somebody can make a script to download malicious content from the Internet.

Pickle should really never be used on any content which is not 100% under the control of the admin of the machine. It really is meant to be used only for temporary caching, for serializing Python objects for `multiprocessing`, and similar very internal purposes, never for storage of the live content or even less for its exchange.

That’s the reason I suggested so [harsh solution as above](https://github.com/nltk/nltk/issues/3266#issuecomment-2204404662).

**Comment by nicolaschaillan** [comment]
> > The solution is simple. Whitelist which pickles can be loaded. Give control to the tenant to decide which ones are fine or not. Maybe by name or something. Right now, the fear is that third party can manage to load some that are malicious.
> 
> I don’t think, that’s good enough. As long as there is an opportunity how to make NLTK download pickle from the Internet (by using some internal undocumented functions, there are really no private and really inaccessible functions in Python), somebody can make a script to download malicious content from the Internet.
> 
> Pickle should really never be used on any content which is not 100% under the control of the admin of the machine. It really is meant to be used only for temporary caching, for serializing Python objects for `multiprocessing`, and similar very internal purposes, never for storage of the live content or even less for its exchange.
> 
> That’s the reason I suggested so [harsh solution as above](https://github.com/nltk/nltk/issues/3266#issuecomment-2204404662).

To be clear, I was proposing to have a whitelist feature to block the pickle loading function from loading any pickle that wouldn't be whitelisted, that would also solve the internet problem you're mentioning.

**Comment by alvations** [comment]
Please send an email with the CVE URL to me (<username>@gmail.com) and also the email in security.md

I can take a look at it for the next few days

**Comment by adivekar-utexas** [comment]
I think whitelist/allowlist is not that useful, since it's a bit cumbersome to maintain. It becomes a file in the user's home directory, which hurts code portability (you need the file everywhere the code is run).

[HuggingFace in 2022 faced a similar remote code execution vulnerability](https://www.youtube.com/watch?v=2ethDz9KnLk). Their solution was in two parts:
1. Add a flag `trust_remote_code=True` at the  model/tokenizer level, e.g. `AutoModelForCausalLM.from_pretrained('gpt2-xl', ..., trust_remote_code=True)` 
2. Add a `TRUST_REMOTE_CODE` environment variable (defaulting to False), which allowed remote code execution for all modules. [Here is an example](https://huggingface.co/docs/transformers/main/en/model_doc/transfo-xl#transformer-xl)

---

I feel like (1) is a better long-term fix, but it seems like it could take a while to implement, given the number of modules. 

@ekaf @stevenbird does (2) seem like a good idea and doable? 

Concrete proposal: A single env variable `TRUST_REMOTE_NLTK_CODE`, which the user must manually enable in their code via `os.environ["TRUST_REMOTE_NLTK_CODE"] = "True"`, prior to calling `nltk` modules.
- If enabled, then module pickles can be downloaded as usual.
- If not enabled, then remote code download/execution is blocked, using something like [the patch suggested](https://github.com/nltk/nltk/issues/3266#issuecomment-2204404662) by @mcepl. Already-downloaded code modules should be executable (but maybe you want to introduce a `STRICT_*` flag to control this).

I feel like this is a good solution because it gives users control on when to execute remote code, without changing more than 1 line of code. 
- Experimentation code (in Jupyter notebooks) will probably be okay to set `TRUST_REMOTE_NLTK_CODE=True`.
- Production code (running in Docker/etc) will probably pre-download the module pickles during the build phase, but run live using `TRUST_REMOTE_NLTK_CODE=False`.

Let me know what you think.

**Comment by ekaf** [comment]
@alvations, it is great to hear that you believe in the "deep clean" solution, where the pickles are completely removed forever.

There is also excellent news from gpt-4o, suggesting to serialize the more complex data types using Python's _joblib_ module: 
[chatgpt-4o-remedy.txt](https://github.com/user-attachments/files/16105434/chatgpt-4o-remedy.txt)
That is much easier than previously believed, and I have verified that the serialization works on the three complex pickle classes used with _puntkt_, _maxent\_ne\_chunker_, and _maxent\_treebank\_pos\_tagger_.
We need to verify that adjusted  loaders can work as expected with the joblib dumps, and if everything goes well, it seems an ETA is within close reach now.

**Comment by Dunedan** [comment]
> There is also excellent news from gpt-4o, suggesting to serialize the more complex data types using Python's joblib module

`joblib` uses pickle under the hood as well, so this wouldn't solve the issue.

**Comment by ekaf** [comment]
> joblib uses pickle under the hood as well, so this wouldn't solve the issue.

Thanks @Dunedan! Instead, there seems to be a possibility to translate the pickles into Protobuf messages, but it may not be the most convenient, so I guess a combination of Json and .py code would be better.

**Comment by nicolaschaillan** [comment]
https://github.com/advisories/GHSA-cgvx-9447-vcch

**Comment by CSi-CJ** [comment]
https://nvd.nist.gov/vuln/detail/CVE-2024-39705. Hi, Is there an ETA?
NVD doesn’t care how we use it, and doesn’t care how we avoid risks. They only care that there may be risks if we use it.

**Comment by jdwhitaker** [comment]
@Dunedan I see the CVSS base metrics says "Privileges required: None". I believe that's inaccurate and has elevated the CVSS score beyond what it should be.

You mentioned 2 ways this could be exploited, and they both seem to require privileges.
- Controlling the NLTK data index presumably requires privileges.
- Editing pickles in transit as a MITM presumably requires breaking TLS1.3, which requires knowledge of a privileged cryptographic key. If your POC was able to successfully edit pickles in transit, then that's a vulnerability that extends beyond just using pickles to insufficient encryption of data in transit.

**Comment by sandimciin** [comment]
> Having a built-in downloader makes it harder than IMHO necessary to depend on nltk. I understand that some dependencies are several gigabytes in size, but even that should be manageable with a package manager like poetry/nix reusing a local download.
> 
> My vote would be to optionally allow depending on individual nltk data as packages, so that I can e.g. depend on `nltk-punkt-english` and know that a certain version of this data has been "locked" in my poetry lockfile.

https://github.com/nltk/nltk/issues/2228#issuecomment-1648551210

**Comment by Dunedan** [comment]
> @Dunedan I see the CVSS base metrics says "Privileges required: None". I believe that's inaccurate and has elevated the CVSS score beyond what it should be.

I wasn't involved in calculating the severity. I believe that was done by Github's Security Curators team. If you believe that metric isn't accurate, there is a link on the page to suggest improvements for the vulnerability details.

That said, from my perspective "Privileges required: None" seems to be correct, as it refers to privileges on the targeted system prior to the attack. See https://www.first.org/cvss/v4.0/specification-document#Privileges-Required-PR for details.

**Comment by ekaf** [comment]
@Dunedan,  to tamper with a user's files, the attacker needs at least the same privileges as that user. Then the definition of "Privileges Required: Low" seems to apply:

> The attacker requires privileges that provide basic capabilities that are typically limited to settings and resources owned by a single low-privileged user.

According to that definition, @jdwhitaker's [interpretation](https://github.com/nltk/nltk/issues/3266#issuecomment-2214759244l) seems correct.

**Comment by digi604** [comment]
According to our SLA, we would like to have a fix for this in the next 7 days. Otherwise, we are forced to rip out NLTK

**Comment by Dunedan** [comment]
> @Dunedan, to tamper with a user's files, the attacker needs at least the same privileges as that user.

The attacker gains these permissions as part of the attack. What the "Privileges required" metric is about are privileges the attacker needs *prior* to the attack. For this vulnerability an attacker doesn't need any privileges on a targeted system prior to an attack.

**Comment by cbornet** [comment]
> According to our SLA, we would like to have a fix for this in the next 7 days. Otherwise, we are forced to rip out NLTK

How much are you willing to pay to have a fix in less than 7 days ?

**Comment by MichaelMcAleer** [comment]
> According to our SLA, we would like to have a fix for this in the next 7 days. Otherwise, we are forced to rip out NLTK

Investigating removal and replacement today, not a level of risk we are willing to accept.

**Comment by lucasgadams** [comment]
Am I right in my understanding that if you are not explicitly running nltk.download then you are not at risk of this vulnerability? When we included nltk in our codebase a year ago, we did a one time download of the tokenizers we needed and then stored that in a location we control. We point NLTK to the data directory using the NLTK_DATA environment variable. If we assume that there has not been any exploitation of this vulnerability yet and the pickles from a year ago are safe, then I should be able to accept this risk? Is there a way to validate whether or not the existing pickles are safe and do not contain malicious code? And then checksum against known secure versions? Not talking about an actual fix just wondering whether I can suppress my vulnerability alert while we wait for a better fix.

**Comment by ayushxx7** [comment]
following thread...

**Comment by ekaf** [comment]
There are now PRs available, with fixed data packages and nltk handlers for all the suspect packages in _nltk_data_. So an ETA for the fix should be near.
Meanwhile, it could be useful if more users would review the PRs,
and eventually discuss alternatives. In particular, it would be helpful to confirm that _nltk.data_ loads Json and Yaml data in a safe way.

**Comment by noren95** [comment]
@ekaf Thank you for the update!
Is this the fixing PR? https://github.com/nltk/nltk/pull/3286/
and is there an expected 3.8.2 release of nltk?

**Comment by ekaf** [comment]
Out of 6 data packages containing pickles, 3 have been replaced by Json and their new handlers were already  merged into the _develop_ branch last week:

- help/tagsets
- taggers/averaged_perceptron_tagger
- taggers/averaged_perceptron_tagger_ru

The following have been replaced by Tab-files and are awaiting review:

- tokenizers/punkt (#3283)
- chunkers/maxent_ne_chunker (#3286)
- taggers/maxent_treebank_pos_tagger

The last [package is fixed](https://github.com/nltk/nltk_data/pull/218), and #3286 now also includes a handler for it, although that package was not used from nltk before.

A release is planned, though I don't know about its version number.

**Comment by chetanpadhye** [comment]
As per Aqua scan this is 
CVSS Score NVD CVSSv3 9.8 
NLTK through 3.8.1 allows remote code execution if untrusted packages have pickled Python code, and the integrated data package download functionality is used. This affects, for example, averaged_perceptron_tagger and punkt. 

There is No Running Workloads, When this fix is expected ?

**Comment by chetanpadhye** [comment]
> According to our SLA, we would like to have a fix for this in the next 7 days. Otherwise, we are forced to rip out NLTK

What is alternate you have ?

**Comment by lpi-tn** [comment]
Hello !
I just saw this CVE pop in my alerts,
I download punkt and bcp47, if I stop downloading them from `nltk_data` but use my local file am I avoiding the problem ?

**Comment by mcepl** [comment]
> Out of 6 data packages containing pickles, 3 have been replaced by Json and their new handlers were already merged into the _develop_ branch last week:

Does it mean that the whole pickle opening will be removed from NLTK?

**Comment by cherylking** [comment]
> A release is planned, though I don't know about its version number.

@ekaf Any updates on when we can expect a release that resolves this CVE?

**Comment by cherylking** [comment]
@ekaf An entire week of silence on this issue leaves everyone affected by this CVE scrambling...not knowing whether they have to eliminate the use of this library in their code by the end of July or not. Please respond with an outlook for the release that resolves this CVE.

**Comment by cbornet** [comment]
> @ekaf An entire week of silence on this issue leaves everyone affected by this CVE scrambling...not knowing whether they have to eliminate the use of this library in their code by the end of July or not. Please respond with an outlook for the release that resolves this CVE.

This is OSS so the maintainers don't have to give you an ETA. I don't think they gain a dollar for the great work they do (do you want to be a sponsor ?)
FWIW they could be in holidays for a month or this bug could be never fixed.
If it's not acceptable for you, you can:
* remove the dependency
* live with the CVE and accept the risk (personally I would not download packages at runtime)
* fork and fix it yourself

I'm sure the maintainers do what they can. They know very well that their project reputation is at stake.

**Comment by ekaf** [comment]
Thanks @cbornet, I have also wondered about how people could talk about an SLA when this is free software. 
However, I am sorry for not responding sooner to @cherylking. The reason is that there is no update for the moment, and your guess is as good as mine concerning the release date.
But the outlook seems reasonable: if we can assume that Json is safe, it should be possible to close this issue before the end of July (without any kind of warranty, though).

**Comment by vishwvir-singh** [comment]
@ekaf Will the remote code execution vulnerability of the NLTK cause problems if we are not explicitly using its nltk.download() or load() within the project?
 
In my project, I am using only its regexp_tokenize (`from nltk.tokenize import regexp_tokenize`) module and noting else. Will the NTLK vulnerability still cause any troubles in the production environments?

**Comment by ekaf** [comment]
@vishwvir-singh, please ask someone else. I prefer not to make statements about the future, or assume that anything is safe.

**Comment by mbanon** [comment]
Hello, is the fixed version expected to be available on pyPI sometime soon? Thanks!

**Comment by OliverFarren** [comment]
> Hello, is the fixed version expected to be available on pyPI sometime soon? Thanks!

~~It looks to be being built and run now so yes https://github.com/nltk/nltk/actions/runs/10088968161/job/27895608352~~

As per this https://github.com/nltk/nltk/pull/3290#issuecomment-224465899, release CI will be dependent on #3283

**Comment by nicolaschaillan** [comment]
Can we get this released now without having to wait on #3283?

**Comment by sadra-barikbin** [comment]
Hi, what's the alternative to be able to work with `nltk.sent_toknize()` which does `load(f"tokenizers/punkt/{language}.pickle")` internally?

**Comment by ekaf** [comment]
> Hi, what's the alternative to be able to work with nltk.sent_toknize() which does load(f"tokenizers/punkt/{language}.pickle") internally?

@sadra-barikbin, the alternative is #3283. It redefines _sent_tokenize()_ without pickles. Please test it if you can and report any problems.

**Comment by nikky78** [comment]
Any news on this ?

**Comment by acompa** [comment]
Checking in, as #3283 seems to be approved for a merge as of yesterday, and #3290 has merged.

Also noticing this issue is closed, which seems incorrect since there is no public release of `nltk` which fully addresses this vulnerability.

**Comment by nikky78** [comment]
On Pypi, it's still the 3.8.1 version. Does the nltk.download works again ?

**Comment by nikky78** [comment]
Is there a way to download the "averaged_perceptron_tagger" from nltk now please ?

**Comment by ekaf** [comment]
@nikky78, the most convenient is to use NLTK's high-level _pos_tag_ function, which is unchanged:

`from nltk.tag import pos_tag`

But you can also call the PerceptronTagger explicitly:

```
from nltk.tag import PerceptronTagger
tagger = PerceptronTagger()
```

The same also applies to [the other replaced packages](https://github.com/nltk/nltk/pull/3283#issuecomment-2264705572).

**Comment by nikky78** [comment]
> @nikky78, the most convenient is to use NLTK's high-level _pos_tag_ function, which is unchanged:
> 
> `from nltk.tag import pos_tag`
> 
> But you can also call the PerceptronTagger explicitly:
> 
> ```
> from nltk.tag import PerceptronTagger
> tagger = PerceptronTagger()
> ```
> 
> The same also applies to [the other replaced packages](https://github.com/nltk/nltk/pull/3283#issuecomment-2264705572).

I'm trying to use langchain UnstructuredMarkdownLoader. But it requires nltk averaged_perceptron_tagger. So I can't use your way I believe.

**Comment by ekaf** [comment]
@nikky78, there should be no difference between the old and the new averaged perceptron tagger object. Please read the code, and report actual differences, instead of vague beliefs.

**Comment by nikky78** [comment]
> @nikky78, there should be no difference between the old and the new averaged perceptron tagger object. Please read the code, and report actual differences, instead of vague beliefs.

I do :
loader = UnstructuredMarkdownLoader(markdown_path)
documents = loader.load()

It does not seem to be take a tagger in arguments

**Comment by rejuls** [comment]
It looks like this and the related PRs are merged. But the fix has not yet reached pypi. pypi is still 3.8.1. Could this be released as 3.8.2 or are there any blockers? @stevenbird @ekaf ?

**Comment by ekaf** [comment]
Thanks @noren95,actually #3283 still needs to be merged.

**Comment by ekaf** [comment]
Now that the last PR is merged, a complete pre-release zip is available from Github, using the "Code" button near the top of the main page.
[This document](https://github.com/nltk/nltk/blob/develop/RELEASE-HOWTO.txt) lists what needs to be done for an official release. While @stevenbird is out doing fieldwork without enough bandwidth, let's hope @purificant will manage it.

**Comment by stevenbird** [comment]
I can do it on Friday.

**Comment by ekaf** [comment]
Meanwhile, you can already  [dowload the latest _develop_ zip of NLTK](https://github.com/nltk/nltk/archive/refs/heads/develop.zip).

**Comment by cassneal** [comment]
@stevenbird thank you! we are keeping an eye out for the new release.

**Comment by ekaf** [comment]
Now that NLTK 3.8.2 is released, users need definitively to upgrade. Those who are not yet convinced that this vulnerability is highly critical will soon be, if they have a look at [this detailed analysis](https://www.vicarius.io/vsociety/posts/rce-in-python-nltk-cve-2024-39705-39706). 
It contains a full Proof-of-Concept of the attack, which is specially scary for those who are using systems that allow you by default to _sudo_ tasks without asking for the root password. It sounds weird, but such systems are actually common.

**Comment by cbornet** [comment]
Can the CVE be informed that 3.8.2 fixes the vulnerability so checkers are not triggered anymore ?

**Comment by ekaf** [comment]
> Can the CVE be informed that 3.8.2 fixes the vulnerability so checkers are not triggered anymore ?

Certainly, but how do you inform the CVE?

**Comment by cbornet** [comment]
No idea 😅

**Comment by dbIgel** [comment]
You can directly edit/"improve" it at https://github.com/advisories/ghsa-cgvx-9447-vcch/
(which would be https://github.com/advisories/ghsa-cgvx-9447-vcch/improve )

**Comment by cbornet** [comment]
Actually the conf in the NVD is `"lessThanOrEqual":"3.8.1"` so checkers are OK now.

**Comment by jdwhitaker** [comment]
> Now that NLTK 3.8.2 is released, users need definitively to upgrade. Those who are not yet convinced that this vulnerability is highly critical will soon be, if they have a look at [this detailed analysis](https://www.vicarius.io/vsociety/posts/rce-in-python-nltk-cve-2024-39705-39706). It contains a full Proof-of-Concept of the attack, which is specially scary for those who are using systems that allow you by default to _sudo_ tasks without asking for the root password. It sounds weird, but such systems are actually common.

The lack of HTTPS takes the vulnerability from “bad practice” to “lethal”

**Comment by nicolaschaillan** [comment]
You need to do it on the NVD website and add a CPE at the bottom so it shows the latest version as fixed.

https://nvd.nist.gov/vuln/detail/CVE-2021-43854

**Comment by ekaf** [comment]
> The lack of HTTPS takes the vulnerability from “bad practice” to “lethal”

Thanks @jdwhitaker, #3295 should ensure that the downloader only uses HTTPS.

**Comment by sushant-glean** [comment]
Hi @ekaf, the fixed package (3.8.2) was on [pypi](https://pypi.org/project/nltk/) yesterday, seems like it has been removed now. Any reason for the same?

**Comment by nicolaschaillan** [comment]
This needs to be reopened...

**Comment by dbIgel** [comment]
The "fix"-release is now 3.9b1, see https://github.com/nltk/nltk/issues/3301#issuecomment-2292549011 for reference.

**Comment by cbornet** [comment]
> The "fix"-release is now 3.9b1, see https://github.com/nltk/nltk/issues/3301#issuecomment-2292549011 for reference.

Would it be possible to cut a "non-beta" release ? 3.9.0 ?

**Comment by stevenbird** [comment]
Please see 3.9.1

**Comment by InterferencePattern** [comment]
For 3.9.1 to address the NIST CVE, it would mean that the download of unsafe pickles is no longer possible, right? If that's the case, it wouldn't be necessary to start using `nltk.download("punkt_tab")`, right? 

`nltk.download("punkt")` should now be pulling `punkt_tab` instead?

I see many projects bumping the version to 3.9.1 _and_ changing the downloaded module name to "punkt_tab"

**Comment by ekaf** [comment]
@InterferencePattern, the package switching is only implemented for the _load()_ method. But you need to download('punkt_tab').

**Comment by cornzz** [comment]
Hi @ekaf,
sorry to ask again but I am unclear on this... if I am simply using [`sent_tokenize()`](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/__init__.py#L109) without explicitly downloading anything, do I have to take action too if I am on 3.8.1? That function also uses the punkt tokenizer but I couldn't quite understand if it also downloads a pickle when initializing the [`PunktTokenizer`](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L1737) and is therefore affected by this vulnerability...

Also, what do I have to pay attention to when moving to 3.9.1, are there breaking changes? Do I have to change `nltk.download("punkt")`? There are no necessary migration steps outlined on https://www.nltk.org/news.html, but it does seem from your last comment I would have to change the call to `nltk.download("punkt_tab")`?
It would be helpful to have a migration guide outlining what has to be done moving from 3.8.1 to 3.9.1.

Thanks for all your work

**Comment by ekaf** [comment]
Yes @cornzz, sent_tokenize() used to rely on a pickle, and was thus vulnerable to a potentially lethal attack. So nobody should be on 3.8.1. You need to upgrade to NLTK 3.9.1. 

Concerning a migration guide, if you use high-level functions nothing changes, [as shown in these examples](https://github.com/nltk/nltk/pull/3283#issuecomment-2264705572).
What changes is that users who want more low-level control, no longer call _load_ directly, but now import the corresponding class instead.
For examples of that, please have a look at how the switch\_* functions in _data.py_ are called (L818-829) and implemented (L667-731).

**Comment by nikky78** [comment]
I would like to use langchain UnstructuredMarkdownLoader like this:
```
from langchain.document_loaders import UnstructuredMarkdownLoader
loader = UnstructuredMarkdownLoader(markdown_path)
documents = loader.load()
```
But I get the error:
```
  Resource punkt_tab not found.
  Please use the NLTK Downloader to obtain the resource:
```

Is there a way with the new version to do nltk.download('punkt_tab') ?

**Comment by wallies** [comment]
> I would like to use langchain UnstructuredMarkdownLoader like this:
> 
> ```
> from langchain.document_loaders import UnstructuredMarkdownLoader
> loader = UnstructuredMarkdownLoader(markdown_path)
> documents = loader.load()
> ```
> 
> But I get the error:
> 
> ```
>   Resource punkt_tab not found.
>   Please use the NLTK Downloader to obtain the resource:
> ```
> 
> Is there a way with the new version to do nltk.download('punkt_tab') ?

@nikky78 
Did you get a similar error mentioned here https://github.com/langchain-ai/langchain/issues/25609

nltk.download('punkt_tab') should work.

**Comment by nikky78** [comment]
> > I would like to use langchain UnstructuredMarkdownLoader like this:
> > ```
> > from langchain.document_loaders import UnstructuredMarkdownLoader
> > loader = UnstructuredMarkdownLoader(markdown_path)
> > documents = loader.load()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But I get the error:
> > ```
> >   Resource punkt_tab not found.
> >   Please use the NLTK Downloader to obtain the resource:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is there a way with the new version to do nltk.download('punkt_tab') ?
> 
> @nikky78 Did you get a similar error mentioned here [langchain-ai/langchain#25609](https://github.com/langchain-ai/langchain/issues/25609)
> 
> nltk.download('punkt_tab') should work.

Indeed, it worked ! thank ^^

**Comment by petfik** [comment]
Sharing my upgrade from 3.8.1 to 3.9.1, tokenizer use case:

NLTK 3.8.1:
```python
# download data
import nltk
nltk.download('punkt')

# put pickles to a local folder manually

# at runtime
import nltk
file_name = f'{NLTK_MODEL_PATH}/{lang}.pickle'  # lang e.g. 'english'
tokenizer = nltk.data.load(file_name)
tokenizer.tokenize(text)
```

NLTK 3.9.1:
```python
# download data
import nltk
nltk.download('punkt_tab')

# put the nltk_data folder contents (with folder structure preserved) to a local folder
# e.g. local_path/tokenizers/punkt_tab/english/<four files here>

# at runtime
import nltk
from nltk.tokenize import PunktTokenizer
nltk.data.path.append(NLTK_MODEL_PATH)
tokenizer = PunktTokenizer(lang)  # lang e.g. 'english'
tokenizer.tokenize(text)
```

**Comment by FilippTrigub** [comment]
Hello! I've upgrade to 3.9.1, but still cant complete the download. Please suggest, what I might be missing!

![image](https://github.com/user-attachments/assets/2575d2d2-3a55-4573-a5bb-a175c89010f2)

**Comment by davidanhip** [comment]
I'm on version 3.9.1 and I got confused because the .load method was no longer in PerceptronTagger.  If the .load method isn't available, should the load=False flag still be supported in the __init__?  I'm not sure if there are other use cases, but if calling PerceptronTagger(load=False) results in a class that will never be of use, the load flag should also be removed.
